src/human_player_module.c

```diff
@@ -0,0 +1,391 @@
+    {
+        request_card(HUMAN, 1);
+    }
+
+    else if (user_input[0] == 'e' || user_input[0] == 'E')
+    {
+        if (g_card_requested)
+        {
+            end_turn(g_player_on_turn);
+        }
+        else
+        {
+            printf("Please draw a card before you can end your turn");
+            display_player_deck(human_player);
+        }
+    }
+    else
+    {
+        if (validate_card(user_input))
+        {
+            //To be added if the format of user input is different
+            Card_t *human_card_choice = map_user_input(user_input);
+            //If valid add the card to the discard_pile
+            if (check_is_valid_turn(&g_card_on_table, human_card_choice))
+            {
+                Deck_t **pp_head = &(human_player->cards_on_hand);
+
+                Card_t *to_remove_card = pick_card_from_deck(pp_head, human_card_choice);
+
+                if (to_remove_card != NULL)
+                {
+                    RET_TYPE(remove_card_from_deck(pp_head, human_card_choice));
+                    add_card_at_end(g_discard_pile, *to_remove_card);
+                }
+            }
+        }
+        else
+        {
+            invalid_turn_warning();
+        }
+    }
+    return SUCCESS;
+}
+/**
+ * @brief Remove a particular card from the given deck
+ * Can be used to discard a particular card - during each turn
+ * Works for both the computer as well as human deck of cards
+ * 
+ * @param pp_head Pointer to the head of the deck from which card has to be removed
+ * @param card_to_be_removed  Card to be removed
+ * @return ret_type_e RET_EMPTY_LIST- If the deck has no cards
+ *                 RET_FAILURE - If the removing of card fails
+ *                 RET_SUCCESS - If the card is been removed successfully
+ */
+ret_type_e remove_card_from_deck(Deck_t **pp_head, Card_t *card_to_be_removed)
+{
+    Deck_t *to_delete = *pp_head;
+    Deck_t *prev = *pp_head;
+    if (*pp_head == NULL)
+    {
+        printf("List is already empty.");
+        return RET_EMPTY_LIST;
+    }
+    while (to_delete->next != NULL)
+    {
+        if (to_delete->card.color == card_to_be_removed->color && to_delete->card.name == card_to_be_removed->name)
+        {
+            prev->next = to_delete->next;
+            free(&(to_delete->card));
+            return SUCCESS;
+        }
+        to_delete = to_delete->next;
+    }
+    return RET_FAILURE;
+}
+
+/**
+ * @brief Function to map the human player input to Card type structure
+ * Should be called only after validation of the user input
+ * 
+ * If user input is 3R, It will be mapped as follows
+ *  card.color=Red, card.name=3
+ * 
+ * @param input Value entered by the human during his turn
+ * 
+ * @return Card_t* Contains the mapped card name and color values
+ * If Card_t is NULL it means the input is an invalid one
+ */
+Card_t *map_user_input(char *input)
+{
+
+    Card_t *temp_card = (Card_t *)malloc(sizeof(Card_t));
+    temp_card = NULL;
```

Comment:
temp_card is a pointer, should not be NULL. Dereferencing a null pointer will typically crash your program.

Reply:
Understood.Thanks!